File: src/acmo/core/services/localization_service.ts
import AsyncStorage from '@react-native-async-storage/async-storage';
import axios, { type AxiosInstance } from 'axios';

const BASE_URL = 'https://api.tyrads.com/v3.0/';

interface TranslationResponse {
  data: Array<{
    code: string;
    sha256: string;
  }>;
}

type Translations = {
  [key: string]: string | Translations;
};

class LocalizationService {
  private static instance: LocalizationService;
  private axios: AxiosInstance | null = null;
  private translations: Translations = {};
  private supportedLocales: string[] = [];
  private readonly fallbackLocale: string = 'en';

  private constructor() {
  }

  public static getInstance(): LocalizationService {
    if (!LocalizationService.instance) {
      LocalizationService.instance = new LocalizationService();
    }
    return LocalizationService.instance;
  }

  private static async getHeadersFromStorage(): Promise<Record<string, string>> {
    try {
      const data = await AsyncStorage.getItem('credentials');
      return data ? JSON.parse(data) : {};
    } catch (error) {
      console.error('Failed to retrieve headers from AsyncStorage', error);
      return {};
    }
  }


  public async init(locale: string): Promise<void> {
    if (this.axios) {
      return;
    }

    const headers = await LocalizationService.getHeadersFromStorage();

    this.axios = axios.create({
      baseURL: BASE_URL,
      timeout: 10000,
      headers: {
        'Content-Type': 'application/json',
        Accept: 'application/json',
        "X-API-Key": headers["X-API-Key"],
        "X-API-Secret": headers["X-API-Secret"],
      }
    });

    this.axios.interceptors.request.use(
      (config) => {
        console.log('====================================');
        console.log(`Request: ${config.method} ${config.baseURL}${config.url}`);
        console.log('Headers:', config.headers);
        console.log('====================================');
        return config;
      },
      (error) => {
        return Promise.reject(error);
      }
    );
    this.axios.interceptors.response.use(
      (res) => {
        console.log('====================================');
        console.log('Response Data:', res.data);
        console.log('====================================');
        return res;
      },
      (error) => {
        return Promise.reject(error);
      }
    );
    await this.loadTranslations(locale);
  }

  private async loadTranslations(locale: string, force = false): Promise<void> {
    const hasUpdate = await this.checkForUpdate(locale, force);

    if (!hasUpdate) {
      const cachedData = await AsyncStorage.getItem(`translations_${locale}`);
      if (cachedData) {
        try {
          this.translations = JSON.parse(cachedData);
          return;
        } catch (e) {
          console.error('Failed to parse cached translations', e);
        }
      }
    }

    await this.fetchTranslations(locale, force);
  }

  private async fetchTranslations(locale: string, force = false): Promise<void> {
    if (!this.axios) return;
    try {
      let effectiveLocale = locale;
      if (!this.supportedLocales.includes(locale)) {
        effectiveLocale = this.fallbackLocale;
      }

      const response = await this.axios.get(
        `translations/${effectiveLocale}`,
        {
          params: {
            force,
            format: 'nested',
          },
        }
      );

      if (response.status === 200) {
        this.translations = response.data as Translations;
        await AsyncStorage.setItem(`translations_${effectiveLocale}`, JSON.stringify(response.data));
      } else {
        console.warn(`Failed to load translations: ${response.status}`);
      }
    } catch (e) {
      if (axios.isAxiosError(e)) {
        console.error('Network error fetching translations:', e.message);
      } else {
        console.error('An unexpected error occurred:', e);
      }
    }
  }

  private async checkForUpdate(locale: string, force = false): Promise<boolean> {
    if (!this.axios) return false;
    try {
      const response = await this.axios.get<TranslationResponse>('translations/version', { params: { force } });
      if (response.status === 200) {
        const data = response.data.data;
        this.supportedLocales = data.map(item => item.code);

        const currentLocaleData = data.find(item => item.code === locale);
        if (!currentLocaleData) {
          return false;
        }

        const currentLocaleSha256 = currentLocaleData.sha256;
        const cachedVersion = await AsyncStorage.getItem(`cached_version_${locale}`);

        if (currentLocaleSha256 !== cachedVersion) {
          await AsyncStorage.setItem(`cached_version_${locale}`, currentLocaleSha256);
          return true;
        }
      }
    } catch (e) {
      if (axios.isAxiosError(e)) {
        console.error('Error checking for update:', e.message);
      } else {
        console.error('An unexpected error occurred:', e);
      }
    }
    return false;
  }

  public translate(key: string, args: Record<string, string | number> = {}): string {
    const keys = key.split('.');
    let currentMap: any = this.translations;

    for (const k of keys) {
      if (typeof currentMap === 'object' && currentMap !== null && currentMap.hasOwnProperty(k)) {
        currentMap = currentMap[k];
      } else {
        return key;
      }
    }

    if (typeof currentMap === 'string') {
      let result = currentMap;
      for (const [argKey, argValue] of Object.entries(args)) {
        const regex = new RegExp(`{${argKey}}`, 'gi');
        result = result.replace(regex, String(argValue));
      }
      return result;
    }

    return key;
  }

  public async changeLanguage(locale: string, force = false): Promise<void> {
    await this.loadTranslations(locale, force);
  }
}

export default LocalizationService;

File: ios/Tyrads/core/utils/Services/LocalizationService.swift
//
//  LocalizationService.swift
//  Pods
//
//  Created by Basharat Mehdi on 02/10/25.
//

import Foundation

final class LocalizationService {
  
  static let shared = LocalizationService()
  private init() {}
  
  private var translations: [String: Any] = [:]
  private var supportedLocales: [String] = []
  
  private let userDefaults = UserDefaults.standard
  private let fallbackLocale = "en"
  
  // MARK: - Public Interface
  func initialize(locale: String) async {
    await loadTranslations(locale: locale)
  }
  
  func changeLanguage(locale: String, force: Bool = false) async {
    await loadTranslations(locale: locale, force: force)
  }
  
  func translate(_ key: String, args: [String: String]? = nil) -> String {
    let keys = key.split(separator: ".").map { String($0) }
    var current: Any? = translations
    
    for k in keys {
      if let dict = current as? [String: Any], let next = dict[k] {
        current = next
      } else {
        return key
      }
    }
    
    guard var result = current as? String else {
      return key
    }
    
    if let args = args {
      for (argKey, argValue) in args {
        let regex = try? NSRegularExpression(pattern: "\\{\(argKey)\\}", options: .caseInsensitive)
        if let regex = regex {
          result = regex.stringByReplacingMatches(in: result, range: NSRange(result.startIndex..., in: result), withTemplate: argValue)
        }
      }
    }
    
    return result
  }
  
  // MARK: - Private Loading Methods
  private func loadTranslations(locale: String, force: Bool = false) async {
    let hasUpdate = await checkForUpdate(locale: locale, force: force)
    
    if !hasUpdate {
      if let cachedData = userDefaults.data(forKey: "translations_\(locale)"),
         let json = try? JSONSerialization.jsonObject(with: cachedData, options: []) as? [String: Any] {
        translations = json
        return
      }
    }
    
    await fetchTranslations(locale: locale, force: force)
  }
  
  private func fetchTranslations(locale: String, force: Bool) async {
    var currentLocale = locale
    if !supportedLocales.contains(currentLocale) {
      currentLocale = fallbackLocale
    }
    
    
    let urlString = "\(AcmoConfig.BASE_URL)translations/\(currentLocale)?force=\(force)&format=nested"
    guard let url = URL(string: urlString) else {
      print("LocalizationService: Invalid URL for translations")
      return
    }
    
    var request = URLRequest(url: url)
    request.httpMethod = "GET"
    request.setTyradsHeaders(with: Tyrads.instance)
    
    do {
      let (data, response) = try await URLSession.shared.data(for: request)
      
      guard let httpResponse = response as? HTTPURLResponse, httpResponse.statusCode == 200 else {
        print("LocalizationService: Failed to load translations: \((response as? HTTPURLResponse)?.statusCode ?? -1)")
        return
      }
      
      if let jsonResponse = try JSONSerialization.jsonObject(with: data, options: []) as? [String: Any] {
        translations = jsonResponse
        
        userDefaults.set(data, forKey: "translations_\(currentLocale)")
      } else {
        print("LocalizationService: Failed to parse JSON response.")
      }
    } catch {
      print("LocalizationService: Network error fetching translations: \(error)")
    }
  }
  
  private func checkForUpdate(locale: String, force: Bool) async -> Bool {
    let urlString = "\(AcmoConfig.BASE_URL)translations/version?force=\(force)"
    guard let url = URL(string: urlString) else {
      print("LocalizationService: Invalid URL for version check")
      return false
    }
    var request = URLRequest(url: url)
    request.httpMethod = "GET"
    request.setTyradsHeaders(with: Tyrads.instance)
    
    do {
      let (data, response) = try await URLSession.shared.data(for: request)
      
      guard (response as? HTTPURLResponse)?.statusCode == 200 else {
        print("LocalizationService: Failed to get version check response.")
        return false
      }
      
      struct VersionData: Decodable {
        let code: String
        let sha256: String
      }
      struct VersionResponse: Decodable {
        let data: [VersionData]
      }
      
      let jsonResponse = try JSONDecoder().decode(VersionResponse.self, from: data)
      let versionData = jsonResponse.data
      
      supportedLocales = versionData.map { $0.code }
      
      guard supportedLocales.contains(locale) else {
        return false
      }
      
      guard let currentLocaleVersion = versionData.first(where: { $0.code == locale }) else {
        return false
      }
      
      let currentSha256 = currentLocaleVersion.sha256
      let cachedVersion = userDefaults.string(forKey: "cached_version_\(locale)")
      
      if currentSha256 != cachedVersion {
        userDefaults.set(currentSha256, forKey: "cached_version_\(locale)")
        return true
      }
      
    } catch {
      print("LocalizationService: Error checking for update: \(error)")
    }
    
    return false
  }
}

extension URLRequest {
  mutating func setTyradsHeaders(with sdk: Tyrads) {
    self.setValue("application/json", forHTTPHeaderField: "Content-Type")
    
    self.setValue(sdk.apiKey, forHTTPHeaderField: "X-API-Key")
    self.setValue(sdk.apiSecret, forHTTPHeaderField: "X-API-Secret")
    
    self.setValue(AcmoConfig.SDK_PLATFORM, forHTTPHeaderField: "X-SDK-Platform")
    self.setValue(AcmoConfig.SDK_VERSION, forHTTPHeaderField: "X-SDK-Version")
  }
}
